src/Microsoft.AspNet.OData/Routing/ODataActionSelector.cs

```diff
@@ -106,5 +112,80 @@ private static HttpResponseMessage CreateErrorResponse(HttpRequestMessage reques
 
             return request.CreateErrorResponse(statusCode, error);
         }
+
+        private static bool ActionParametersMatchRequest(HttpActionDescriptor action, HttpControllerContext context)
+        {
+            var parameters = action.GetParameters();
+            var routeData = context.RouteData;
+            var conventionsStore = context.Request.ODataProperties().RoutingConventionsStore;
+            var matchedBody = false;
+            var route = routeData.Route as ODataRoute;
+            var routePrefix = route?.RoutePrefix;
+            var availableKeys = routeData.Values.Keys
+                .Where(k => routePrefix != "{" + k + "}"
+                    && k != ODataRouteConstants.Action
+                    && k != ODataRouteConstants.Controller
+                    && k != ODataRouteConstants.ODataPath)
+                .Select(k => k.ToUpperInvariant())
+                .ToList();
+
+            if (parameters.Count == 0 && availableKeys.Count > 0)
+            {
+                return false;
+            }
+
+            foreach (var p in parameters)
+            {
+                string parameterName = p.ParameterName.ToUpperInvariant();
+                if (availableKeys.Contains(parameterName))
+                {
+                    continue;
+                }
+                if (conventionsStore.Keys.Any(key => key.Contains(p.ParameterName)))
+                {
+                    continue;
+                }
+                if (context.Request.GetQueryNameValuePairs().Any(kvp => kvp.Key == p.ParameterName))
+                {
+                    continue;
+                }
+                if (!matchedBody && RequestHasBody(context))
+                {
+                    matchedBody = true;
+                    continue;
+                }
```

Comment:
Why is this done inside of the loop? couldn't you do this check before (or after) checking parameters?